server/swap/swap.go

```diff
@@ -260,14 +255,88 @@ func (s *orderSwapTracker) canceled(ord order.Order) {
 
 	// Prevent completion of active swaps from counting the order as
 	// successfully completed.
-	stat.offBook = true
-	stat.hasFailed = true
+	stat.OffBook = true
+	stat.HasFailed = true
+}
+
+type msgAckers struct {
+	Request *msgjson.Message // request to user, may not be nil
+	AckData []*ackerData
+}
+
+type ackerData struct {
+	User    account.AccountID
+	MatchID order.MatchID    // key for a *matchTrackers map
+	Params  msgjson.Signable // msgjson.{Match,Audit,Redemption,RevokeMatch}
+	IsMaker bool
+	//IsAudit bool // inferred if Params.(*msgjson.Audit)
+}
+
+func (s *Swapper) setLiveAckers(req *msgjson.Message, ackers []*messageAcker) {
+	ackDats := make([]*ackerData, len(ackers))
+	for i, ack := range ackers {
+		ackDats[i] = &ackerData{
+			User:    ack.user,
+			MatchID: ack.match.ID(),
+			Params:  ack.params,
+			IsMaker: ack.isMaker,
+		}
+	}
+	s.liveAckersMtx.Lock()
+	s.liveAckers[req.ID] = &msgAckers{
+		Request: req,
+		AckData: ackDats,
+	}
+	s.liveAckersMtx.Unlock()
+}
+
+func (s *Swapper) setLiveAcker(req *msgjson.Message, ackDat *messageAcker) {
+	s.setLiveAckers(req, []*messageAcker{ackDat})
+}
+
+func (s *Swapper) rmLiveAckers(msgID uint64) {
+	s.liveAckersMtx.Lock()
+	delete(s.liveAckers, msgID)
+	s.liveAckersMtx.Unlock()
+}
+
+// Data input to handleInit and handleRedeem, which both start coin waiters with
+// latencyQ. Store handlerArgs before starting the waiter, and remove them when
+// the waiter completes or expires. On Swapper shutdown, the stored handlerArgs
+// should be saved in the state data after stopping the latencyQ.
+type handlerArgs struct {
+	User account.AccountID
+	Msg  *msgjson.Message // init or redeem inferred from Msg.
+}
+
+// client generated message IDs are not globally unique.
```

Comment:
I didn't think this was trouble as the goal of this was just to key uniquely.

But since you mention reconnects, I would expect client-originating requests would have handlers registered by ID that would not be be lost on disconnect (only dexc *shutdown*).  If this association is lost in the client, the client might not know what to do with any responses from `processInit` and `processRedeem` (whether `respondError` or `respondSuccess`).  Off hand I'm don't know what happens to the client's request handler map when the connection goes down, but I thought it was fine as long as the software didn't shutdown.

But otherwise, `processInit` and `processRedeem` don't need the client's msg ID provided by the original `handleInit`/`handleRedeem` to do their jobs, which consist of validating the contract/redeem via the asset backends, storing the validated data, and sending an ack request to the counter party with a fresh server-generated msg ID.  The only possible issue is as I mentioned with the success/fail message sent to the user having an ID that the client no longer recognizes, or even associates with a different request.

With the server-originating requests, I regenerate the msg ID when firing up the live ackers, but it's clearly more of an issue for the client-originating requests.  Do you think this can be handled client side?

Reply:
Yeah, the response handlers map persists across reconnects in client/comms, but of course this is lost on dexc shutdown or anything that makes a fresh `wsConn`.

Reply:
Looking at client/core, it seems that if these coin waiters are live on the server, then either `(*trackedTrade).swapMatches` or `(*trackedTrade).redeemMatches` are sitting at `signAndRequest` waiting to receive and unmarshal a `msgjson.Acknowledgement`.  And it seems `signAndRequest` would not be erroring as long as the request `Send` completed without err (and it would have if the server has has a coin waiter running from it).  The client just gives the connection `time.Minute*5` according to `client/comms.(*wsConn).logReq` to respond.